### PR TITLE
Adding an example on CaiT

### PR DESCRIPTION
**Pull request by sayakpaul**
This PR adds a tutorial on the CaiT model proposed in [Going deeper with Image Transformers](https://arxiv.org/abs/2103.17239).

This tutorial slightly deviates from the usual architecture implementation tutorials. Specifically, I have intentionally kept the training part out of the notebook to focus more on the implementation and visualization aspects. 

The readers can always refer to the notebooks I have linked from the tutorial in case they're interested in the training part. 

Cc: @fchollet

### CODE REVIEW

File: examples/vision/cait.py [hunk 1]
```diff
@@ -0,0 +1,860 @@
...
+
+# ImageNet-1k class labels.
+imagenet_labels = (
+    "https://storage.googleapis.com/bit_models/ilsvrc2012_wordnet_lemmas.txt"
+)
+label_path = tf.keras.utils.get_file(origin=imagenet_labels)
+
+with open(label_path, "r") as f:
+    lines = f.readlines()
+imagenet_int_to_str = [line.rstrip() for line in lines]
```
Comment (fchollet) [inline examples/vision/cait.py:679]: The name sounds like a dict but it's a list

File: examples/vision/cait.py [hunk 2]
```diff
@@ -0,0 +1,860 @@
...
+
+"""
+## Visualizing the Attention Layers
+
+We start by inspecting the shape of the attention weights returned by a Class Attention
+layer.
+"""
+
+# (batch_size, nb_attention_heads, num_cls_token, seq_length)
+print(ca_atn_score_dict["ca_ffn_block_0_att"].shape)
```
Comment (fchollet) [inline examples/vision/cait.py:718]: Please format the print call to provide context on what is being printed.

File: examples/vision/cait.py [hunk 3]
```diff
@@ -0,0 +1,860 @@
...
+        attentions = np.expand_dims(attentions, -1)
+        print(attentions.shape)
+
+    # Resize the attention patches to 224x224 (224: 14x16)
+    attentions = tf.image.resize(
+        attentions,
+        size=(h_featmap * patch_size, w_featmap * patch_size),
+        method="bicubic",
+    )
+    print(attentions.shape)
```
Comment (fchollet) [inline examples/vision/cait.py:781]: Please remove debugging prints

File: examples/vision/cait.py [hunk 4]
```diff
@@ -0,0 +1,860 @@
...
+        self.scale = head_dim**-0.5
+
+        self.q = keras.layers.Dense(config.projection_dim)
+        self.k = keras.layers.Dense(config.projection_dim)
+        self.v = keras.layers.Dense(config.projection_dim)
+        self.attn_drop = keras.layers.Dropout(config.dropout_rate)
+        self.proj = keras.layers.Dense(config.projection_dim)
+        self.proj_drop = keras.layers.Dropout(config.dropout_rate)
+
+    def call(self, x, training):
```
Comment (fchollet) [inline examples/vision/cait.py:201]: Please provide a default (False) for `training`.

File: examples/vision/cait.py [hunk 5]
```diff
@@ -0,0 +1,860 @@
...
+        B, N, C = tf.shape(x)[0], tf.shape(x)[1], tf.shape(x)[2]
```
Comment (fchollet) [inline examples/vision/cait.py:202]: Please use lowercase for variable names.

File: examples/vision/cait.py [hunk 6]
```diff
@@ -0,0 +1,860 @@
...
+        self.attn_drop = keras.layers.Dropout(config.dropout_rate)
+
+        self.proj = keras.layers.Dense(config.projection_dim)
+
+        self.proj_l = keras.layers.Dense(config.num_heads)
+        self.proj_w = keras.layers.Dense(config.num_heads)
+
+        self.proj_drop = keras.layers.Dropout(config.dropout_rate)
+
+    def call(self, x, training):
```
Comment (fchollet) [inline examples/vision/cait.py:276]: Please provide a default value

**Review by fchollet — COMMENTED** [review]
Thank you for the PR. Awesome work! 👍

**Comment by sayakpaul** [comment]
@fchollet up for another pass. Addressed all the comments.

**Comment by sayakpaul** [comment]
@fchollet a gentle ping :)

**Review by fchollet — APPROVED** [review]
Thanks for the updates! Everything LGTM. I pushed some minor copyedits. Please add the generate files.

**Comment by sayakpaul** [comment]
@fchollet added generated files. I had to change the MAX_LOC in `tutobooks.py` to generate the files just as an FYI.

**Comment by fchollet** [comment]
> I had to change the MAX_LOC in tutobooks.py to generate the files just as an FYI.

How many lines did you have?

**Review by fchollet — APPROVED** [review]
Thanks for the great contribution! 👍 
Merging now.

**Comment by sayakpaul** [comment]
> > I had to change the MAX_LOC in tutobooks.py to generate the files just as an FYI.
> 
> How many lines did you have?

424.

**Pull request merged**